rfcs/0000-error-handling.md

```diff
@@ -0,0 +1,153 @@
+
+Less catastrophic errors do have a downside: It's harder to notice that anything is wrong at all. We'll include an error in the graphql response for unhealthy and failed subgraphs. We will improve and document the indexing status API, so that developers can programmatically check details on a subgraph's health and build their own alerts and dashboards.
+
+## Urgency
+
+Sooner is better than later.
+
+## Terminology
+
+A new subgraph health status called "unhealthy" is added, so instead of a subgraph being failed or not, it can be healthy, unhealthy or failed.
+
+## Detailed Design
+
+### Error handling
+
+Subgraphs now have three possible health status: 'Healthy' is normal operation, 'Unhealthy' is syncing with errors, and 'Failed' is halted due to an error.
+
+If a WASM handler terminates with a trap, the subgraph becomes unhealthy. The node discards all entity operations and data source operations done by the handler and continues syncing.
+
+For subgraphs that are not yet synced, any handler error is still a failure and halts syncing, so errors in pending versions will not make their way to the current version. 
+
+When the node starts, we currently reset the failed flag on subgraphs. Failed subgraphs will still be reset to healthy to attempt to recover, but unhealthy subgraphs are kept unhealthy, since the subgraph will already be past the errors.
+
+Any graphql queries to an unhealthy or failed subgraph will include an error of the form:
+
+```
+{
+	message: "Subgraph not healthy"
+}
+```
+
+So that applications will always be aware that queries may have inconsistent data. Developers should refer to the indexing status API for details on the errors.
+
+### Indexing Status API
+
+The indexing status API will be revamped.
+
+The following top level query is added:
+
+    indexingStatusForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+
+Which returns null if the name is not found, and returns the status for the current version of the subgraph otherwise.
+
+A field `lastHealthyBlock`  is added to record the block previous to the first error.
+
+Errors are made more than just a string with the `Error` type.
+
+The complete indexing status API schema is:
+
+```graphql
+type Query {
+  indexingStatusForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+  indexingStatusesForSubgraphName(subgraphName: String!): [SubgraphIndexingStatus!]!
+  indexingStatuses(subgraphs: [String!]): [SubgraphIndexingStatus!]!
```

Comment:
Can't we just have one `indexingStatuses` collection which takes a `subgraphs` and a `names` parameter (both `[String!]`) and returns an indexing status for each entry in those two parameters? Logically, tehre's one collection of statuses that we filter with different criteria (which in this case are implicitly or'd together)

Reply:
Maybe but that doesn't seem like the easiest API to understand and use, I think `indexingStatusForSubgraphName` is exactly what most users want.

Reply:
Right now it is. In the network `indexingStatuses(subgraphs: ["Qm..."])` becomes the more important API.

I agree with keeping names / IDs separate. One thing we may want to add to the `ForSubgraphName` resolvers are `current: Boolean!` and `pending: Boolean!` flags to allow clients to identify the current/pending versions. Right now, I think the order of statuses returned is supposed to communicate which version is current, but I'm not sure it's actually doing that...

Reply:
Ah, my suggestion was to add a `currentVersion` and `pendingVersion` flag to the returned indexing statuses. In the network the current version is determined by what's in the contracts. And pending versions are no longer a thing.

But I think it would be cleaner to add `indexingStatusForCurrentVersion` and `indexingStatusForPendingVersion` for subgraph names now and retire them in the network, like you suggest, and keep the indexing statuses "clear" of non-network features.